files/psql_metrics

```diff
@@ -0,0 +1,598 @@
+  n.nspname || '.' || c.relname || '.' || s.relname,
+  json_build_object(
+    'size_bytes', pg_relation_size(c.oid),
+    'idx_scan', idx_scan,
+    'idx_tup_read', idx_tup_read,
+    'idx_tup_fetch', idx_tup_fetch,
+    'idx_blks_read', idx_blks_read,
+    'idx_blks_hit', idx_blks_hit
+  )
+)
+FROM
+  pg_catalog.pg_class AS c
+JOIN
+  pg_catalog.pg_namespace AS n ON c.relnamespace = n.oid
+JOIN
+  pg_catalog.pg_stat_all_indexes AS s ON c.oid = s.indexrelid
+JOIN
+  pg_catalog.pg_statio_all_indexes AS si ON c.oid = si.indexrelid
+WHERE
+  n.nspname NOT IN ('pg_catalog', 'information_schema')
+  AND c.relkind = 'i'
+  AND pg_relation_size(c.oid) > 16384;
+EOS
+      end
+    end
+
+    # Collect metrics and return as a hash
+    #
+    # @return [Hash] A hash containing various metrics. An `:error`
+    #   key will be present if failures occurred during collection.
+    def to_h
+      return @result unless @result.nil?
+
+      collect_data
+      @result[:error] = @errors unless @errors.empty?
+
+      @result
+    end
+
+    class CLI
+      ARG_SPECS = [['--timeout INT',
+                    Integer,
+                    'Number of seconds to allow for psql invocations to complete.',
+                    'Defaults to 10.'],
+                   ['--output_dir DIR',
+                    'Write metrics to a timestamped file under DIR instead of',
+                    'printing to STDOUT']]
+
+      def initialize(argv = [])
+        @action = :collect_data
+        @options = {debug: false}
+
+        @optparser = OptionParser.new do |parser|
+          parser.banner = 'Usage: psql_metrics [options]'
+
+          parser.on_tail('-h', '--help', 'Show help') do
+            @action = :show_help
+          end
+
+          parser.on_tail('--debug', 'Enable backtraces from errors.') do
+            @options[:debug] = true
+          end
+        end
+
+        store_option = lambda do |hash, key, val|
+          hash[key] = val
+        end
+
+        ARG_SPECS.each do |spec|
+          # TODO: Yell if ARG_SPECS entry contains no --long-flag.
+          long_flag = spec.find {|e| e.start_with?('--')}.split(' ').first
+          option_name = long_flag.sub(/\A-+(?:\[no-\])?/, '').gsub('-', '_').to_sym
+
+          @optparser.on(store_option.curry[@options][option_name], *spec)
+        end
+
+        args = argv.dup
+        @optparser.parse!(args)
+      end
+
+      def run
+        case @action
+        when :show_help
+          $stdout.puts(@optparser.help)
+          return 0
+        end
+
+        # NOTE: A little odd, since tk_metrics uses the certname. But, this
+        #   matches what system_metrics does.
+        hostname = PSQLMetrics::Exec.exec_cmd('/bin/sh', '-c', 'hostname').stdout.strip.gsub('.', '-')
+
+        metrics = PSQLMetrics.new(**@options)
+        data = {servers: {hostname => {postgres: metrics.to_h}}}
+
+        if (output_dir = @options[:output_dir])
+          host_dir = File.join(output_dir, hostname)
```

Comment:
The other collection methods use a `.` separated hostname directory and a `-` separated name in the JSON. This one has `-` for both.

Reply:
After staring at this for a long bit, I think it came from this portion of `system_metrics`:

https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector/blob/master/files/system_metrics#L325

The only reason I can fathom for this behavior is that some timeseries databases, like Graphite, treat `.` as a separator in the name of a metric. So, maybe this `gsub` is trying to sort that issue out?

However, this is the wrong place to do that sort of translation. The hostname is a significant piece of data and should be preserved in its original form. If a script like `json2timeseriesdb` is preparing the output for a database like Graphite, _that script_ should be doing the translation where needed.

I'll drop the `gsub` and make the same change in `vmware_metrics`. I'm going to leave `system_metrics` alone since it has already shipped and these two haven't.

Reply:
ARGH. Never mind, `tk_metrics` also has this behavior:

https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector/blob/master/files/tk_metrics#L21

Ok, I'll leave the substitution and try to correct the directory names :sigh:

Reply:
Thanks for changing this. I am not sure why it was done this way, but it goes back ~4 years to https://github.com/npwalker/pe_metric_curl_cron_jobs/blame/271becea802a347f1ad30b3f76b05e4045883201/templates/puppetdb_metrics.epp#L27

I bet you are right that it was done for processing issues, but it's something we can change in a later release.